twinify/napsu_mq/junction_tree.py

```diff
@@ -68,6 +68,12 @@ def __init__(self, nodes: Set['TreeNode'], edges: Dict[Tuple['TreeNode', 'TreeNo
         self.nodes = nodes
         self.edges = edges
         self.cliques = cliques
+        self.enable_checks = enable_checks
+
+        if self.enable_checks is True:
+            self.check_self()
+
+    def check_self(self) -> None:
         self.init_factor_assignments()
```

Comment:
The calls made in here set a bunch of members of `self` that are never used anywhere else in this class. Checking for external use, at least `self.downward_order` (set in `self.init_node_orders`) is used in `MarkovNetwork::belief_propagation`. While `self.init_node_orders` can be called from another way, it's bad practice to have class members only sometimes set and sometimes uninitialised. So I'd like the following changes:
- things that are only every used by tests should be computed there
- things that are essential members of the class and can be computed at initialisation, should be computed there
- things that are only computable later should be returned from some class functions, but not be class members if at all possible

Given that the tree is not immutable, that suggests to me that external users should just call `compute_downward_order` when they need it and it shouldn't be stored as a class member.
If that computation is so costly that we want to do it once and then cache it as a member, we should think about making trees immutable objects (i.e., `add_edge` and `remove_redundant_nodes` should return new objects instead of mutating `self`)